cmd/agent/main.go

```diff
@@ -57,7 +57,14 @@ func main() {
 			rOpts := new(reporter.Options).InitFromViper(v, logger)
 			tchanBuilder := tchannel.NewBuilder().InitFromViper(v, logger)
 			grpcBuilder := grpc.NewConnBuilder().InitFromViper(v)
-			cp, err := app.CreateCollectorProxy(rOpts, tchanBuilder, grpcBuilder, logger, mFactory)
+			builders := map[reporter.Type]app.ProxyBuilder{}
+			builders[reporter.GRPC] = func(m map[string]string, factory metrics.Factory, logger *zap.Logger) (app.CollectorProxy, error) {
+				return grpc.NewCollectorProxy(grpcBuilder, rOpts.AgentTags, mFactory, logger)
```

Comment:
I would prefer to have a function in each package that matches the `ProxyBuilder` signature, rather than using lambdas here.

Reply:
I have created `proxy_builders.go` in `agent/app` package. They cannot live in the reporter's package bc it creates cyclic imports 

```
# github.com/jaegertracing/jaeger/cmd/agent/app/processors
FAIL	github.com/jaegertracing/jaeger/cmd/agent/app/processors [setup failed]
import cycle not allowed in test
FAIL
package github.com/jaegertracing/jaeger/cmd/agent/app/processors (test)
	imports github.com/jaegertracing/jaeger/tchannel/agent/app/reporter/tchannel
	imports github.com/jaegertracing/jaeger/cmd/agent/app
	imports github.com/jaegertracing/jaeger/cmd/agent/app/processors



GOROOT=/home/ploffay/bin/go #gosetup
GOPATH=/home/ploffay/projects/golang #gosetup
/home/ploffay/bin/go/bin/go test -c -o /tmp/___go_test_github_com_jaegertracing_jaeger_cmd_agent_app github.com/jaegertracing/jaeger/cmd/agent/app #gosetup
FAIL	github.com/jaegertracing/jaeger/cmd/agent/app [setup failed]
# github.com/jaegertracing/jaeger/cmd/agent/app
FAIL
import cycle not allowed in test
package github.com/jaegertracing/jaeger/cmd/agent/app (test)
	imports github.com/jaegertracing/jaeger/tchannel/agent/app/reporter
	imports github.com/jaegertracing/jaeger/cmd/agent/app

```

Reply:
I wasn't suggesting importing the function type, just having functions in tch/grpc packages that match it on the signature.

Reply:
There was also a suggestion to pass pass `Options` instead of listing the parameters. That creates cycle too.

Reply:
oh, but the struct, I see...